src/Lean/Server/CodeActions.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2022 E.W.Ayers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: E.W.Ayers
+-/
+
+import Lean.Server.FileWorker.RequestHandling
+import Lean.Server.InfoUtils
+
+namespace Lean.Server
+
+open Lsp
+open RequestM
+open Snapshots
+
+structure CodeActionData where
+  uri : DocumentUri
+  resolverId : UInt64
+  deriving ToJson, FromJson
+
+def CodeAction.getFileSource! (ca : CodeAction) : DocumentUri :=
+  let r : Except String DocumentUri := do
+    let some data := ca.data?
+      | throw s!"no data param on code action {ca.title}"
+    let data : CodeActionData ← fromJson? data
+    return data.uri
+  match r with
+  | Except.ok uri => uri
+  | Except.error e => panic! e
+
+instance : FileSource CodeAction where
+  fileSource x := CodeAction.getFileSource! x
+
+/-- A code action optionally supporting a lazy code action computation that is only run when the user clicks on
+the code action in the editor.
+
+If you want to use the lazy feature, make sure that the `edit?` field on the `eager` code action result is `none`.
+ -/
+structure LazyCodeAction where
+  /-- This is the initial code action that is sent to the server, to implement -/
+  eager : CodeAction
+  lazy? : Option (IO CodeAction) := none
```

Comment:
```suggestion

The edit should always be computed lazily.
 -/
structure LazyCodeAction where
  /-- Initial code action that is sent to the client, it should not include the edit. -/
  eager : CodeAction
  /-- Final code action, including the edit. -/
  lazy? : Option (IO CodeAction)
```

The more I think about it, I believe we should make lazy edits the default.  We're probably going to have many code actions (particularly in "work-in-progress" files where there are still errors).  I'm still not entirely sure when VS Code requests code actions, but it does request them when you're just moving the cursor.  If we compute them upfront, that would be a lot of code actions to compute and ship.